core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/plangen/LogicalPlan.java

```diff
@@ -103,30 +100,98 @@ public LogicalPlan(
-
         currentOperator.open();
         Schema operatorSchema = currentOperator.getOutputSchema();
         currentOperator.close();
 
         return operatorSchema;
     }
 
+    /**
+     * Updates the current plan and fetch the schema from an operator
+     * @param operatorID, the ID of an operator
+     * @return Schema, which includes the attributes setting of the operator
+     */
+    public Schema getOperatorOutputSchema(String operatorID, Map<String, List<Schema>> inputSchemas)
+            throws PlanGenException, DataflowException {
+
+        IOperator currentOperator = operatorObjectMap.get(operatorID);
+        Schema outputSchema = null;
+        if (currentOperator instanceof ISourceOperator) {
+            outputSchema = currentOperator.transformToOutputSchema();
+        } else if (inputSchemas.containsKey(operatorID)) {
+            List<Schema> inputSchema = inputSchemas.get(operatorID);
+            try {
+                outputSchema = currentOperator.transformToOutputSchema(
+                        inputSchema.toArray(new Schema[inputSchema.size()]));
+            } catch (TexeraException e) {
+                System.out.println(e.getMessage());
+            }
+        }
+        return outputSchema;
+    }
+
+    /**
+     * For each operator, get its input schema based on the topological order of the graph
+     * @return Map where id of the operator as the key and input schema as the value
+     * @throws PlanGenException
+     */
+    public Map<String, List<Schema>> retrieveAllOperatorInputSchema() throws PlanGenException {
+
+        buildOperators();
+        checkGraphCyclicity();
+
+        // Calculate the in-edge count of each operator
+        Map<String, Integer> inEdgeCount = new HashMap<>();
+        for (Map.Entry<String, LinkedHashSet<String>> entry: adjacencyList.entrySet()) {
+            inEdgeCount.putIfAbsent(entry.getKey(), 0);
+            for (String to: entry.getValue()) {
+                inEdgeCount.put(to, inEdgeCount.getOrDefault(to, 0)+1);
+            }
+        }
+
+        // Get all the source operator
+        Queue<String> queue = new LinkedList<>();
+        for (Map.Entry<String, IOperator> entry: operatorObjectMap.entrySet()) {
+            if (entry.getValue() instanceof ISourceOperator) {
+                queue.add(entry.getKey());
+            }
+        }
+
+        // Retrieve input schema based on topological order
+        Map<String, List<Schema>> inputSchemas = new HashMap<>();
+        while (!queue.isEmpty()) {
+            String origin = queue.poll();
+            Schema curOutputSchema = getOperatorOutputSchema(origin, inputSchemas);
+
+            if (curOutputSchema != null) {
+                for (String destination: adjacencyList.get(origin)) {
+                    if (!(operatorObjectMap.get(destination) instanceof ISink)) {
+                        inputSchemas.computeIfAbsent(destination, k -> new ArrayList<>()).add(curOutputSchema);
```

Comment:
I don't think `computeIfAbsent` is used properly here. If I understand the docs correctly, `computeIfAbsent` is to put a value into a map if the key is not in the map. If the key is already in the map, then the old value is not changed, and in our case, the new outputSchema is simply ignored and thrown away.

Is this really the behavior we wanted? I know that currently we are primarily using single-input operators, and this scenario will only happen if the operator has 2 or more input operators. But this is likely to cause a bug in the future. 

The logic should be:
```Java
if (inputSchemas.containsKey(destination)) {
    inputSchemas.get(destination).add(currentOutputSchema);
} else {
    inputSchemas.put(destination, new ArrayList<>(Arrays.asList(currentOutputSchema)));
}
```
Or in Java 8:
```Java
inputSchemas.merge(destination, Arrays.asList(currentOutputSchema), 
    (list1, list2) -> Streams.concat(list1.stream(), list2.stream()).collect(Collectors.toList()));
```

@avinash0161 this problem needs your attention